### DISCUSSION TITLE
How do I search for PRs with conflicts?

### DISCUSSION THREAD

**Discussion by aaronfranke** [community/community#23152] [New to GitHub]
<p>I am trying to filter search results for PRs with conflicting files such that it says this in the PR:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/c/a/ca889ba8e295d79121f095f1d0c1b0f6ba066f7e.png" data-download-href="/uploads/short-url/sTH85ksRnKLVzRJDDk903YXqW62.png?dl=1" title="Screenshot from 2020-10-12 09-10-34" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/1646875/181075362-bfedd950-9254-4b0f-ad65-7e60151b6351.png" alt="Screenshot from 2020-10-12 09-10-34" data-base62-sha1="sTH85ksRnKLVzRJDDk903YXqW62" width="690" height="162" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/c/a/ca889ba8e295d79121f095f1d0c1b0f6ba066f7e_2_690x162.png, https://aws1.discourse-cdn.com/github/original/3X/c/a/ca889ba8e295d79121f095f1d0c1b0f6ba066f7e.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/c/a/ca889ba8e295d79121f095f1d0c1b0f6ba066f7e.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/c/a/ca889ba8e295d79121f095f1d0c1b0f6ba066f7e_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot from 2020-10-12 09-10-34</span><span class="informations">881×208 12 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>There doesn’t seem to be any information on this in the documentation:</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/free-pro-team@latest/github/searching-for-information-on-github/searching-issues-and-pull-requests#search-by-missing-metadata" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="690" height="162">

<h3><a href="https://docs.github.com/en/free-pro-team@latest/github/searching-for-information-on-github/searching-issues-and-pull-requests#search-by-missing-metadata" target="_blank" rel="noopener nofollow ugc">Searching issues and pull requests - GitHub Docs</a></h3>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by francisfuzz** [comment, marked as answer]
<p><a class="mention" href="/aaronfranke">@aaronfranke</a> - Great question! There’s not a specific qualifier in our search interface that would allow you to fetch a list of pull requests and filtering them for those with conflicting files.</p>
<p>One alternative approach that might be of interest is using our GraphQL API to fetch a list of pull requests and obtaining their number, URL and <code>mergeable</code> value. The <a href="https://docs.github.com/en/free-pro-team@latest/graphql/reference/enums#mergeablestate"><code>mergeable</code></a> field can be one of three values, where <code>CONFLICTING</code> would indicate that the pull request can’t be merged due to merge conflicts.</p>
<p>Here’s an example using the <code>github/linguist</code> repository:</p>
<pre><code class="lang-auto">query {
  repository(owner:"github", name:"linguist") {
    pullRequests(last: 10) {
      edges {
        node {
          number
          url 
          mergeable
        }
      }
    }
  }
}
</code></pre>
<p>Here’s the result set at this time of writing:</p>
<pre><code class="lang-auto">{
  "data": {
    "repository": {
      "pullRequests": {
        "edges": [
          {
            "node": {
              "number": 5029,
              "url": "https://github.com/github/linguist/pull/5029",
              "mergeable": "MERGEABLE"
            }
          },
          {
            "node": {
              "number": 5030,
              "url": "https://github.com/github/linguist/pull/5030",
              "mergeable": "MERGEABLE"
            }
          },
          {
            "node": {
              "number": 5034,
              "url": "https://github.com/github/linguist/pull/5034",
              "mergeable": "UNKNOWN"
            }
          },
          {
            "node": {
              "number": 5035,
              "url": "https://github.com/github/linguist/pull/5035",
              "mergeable": "MERGEABLE"
            }
          },
          {
            "node": {
              "number": 5036,
              "url": "https://github.com/github/linguist/pull/5036",
              "mergeable": "UNKNOWN"
            }
          },
          {
            "node": {
              "number": 5037,
              "url": "https://github.com/github/linguist/pull/5037",
              "mergeable": "UNKNOWN"
            }
          },
          {
            "node": {
              "number": 5040,
              "url": "https://github.com/github/linguist/pull/5040",
              "mergeable": "MERGEABLE"
            }
          },
          {
            "node": {
              "number": 5049,
              "url": "https://github.com/github/linguist/pull/5049",
              "mergeable": "UNKNOWN"
            }
          },
          {
            "node": {
              "number": 5050,
              "url": "https://github.com/github/linguist/pull/5050",
              "mergeable": "CONFLICTING"
            }
          },
          {
            "node": {
              "number": 5051,
              "url": "https://github.com/github/linguist/pull/5051",
              "mergeable": "CONFLICTING"
            }
          }
        ]
      }
    }
  }
}
</code></pre>
<p>You could adapt this approach for a different repository or a set of repositories scoped under an account or organization. I highly recommend checking out our <a href="https://docs.github.com/en/free-pro-team@latest/graphql">documentation covering the GitHub GraphQL API</a>, in addition to our <a href="https://docs.github.com/en/free-pro-team@latest/graphql/guides">guides</a> if you haven’t already. I hope this helps!</p>

**Comment by ferdnyc** [comment]
<p>There are now <a href="https://github.com/marketplace?type=actions&amp;query=conflict" rel="noopener nofollow ugc">several actions</a> in the Github Marketplace that will check for PRs with conflicts and add a label to them, allowing the list to easily be filtered by that label. I think that’s the route I’m going to take.</p>
<p>(Though it sure would be nice if Github’s PR list automatically displayed an icon for conflicting PRs. The same way it does for PRs with linked issues, and for issues with linked PRs.)</p>

**Comment by ferdnyc** [comment]
<p>Took a little auditioning and experimentation, but I now have a Github Actions workflow for <a href="https://github.com/eps1lon/actions-label-merge-conflict" rel="noopener nofollow ugc">eps1lon/actions-label-merge-conflict</a> installed in both of our project repos that receive regular PR contributions.</p>
<p>So far it seems to be working out well, every time there’s a push to our default branch, it scans all of the open PRs for detected conflicts, adds a label of my choosing, and leaves a comment on the PR (to generate an email notification to the submitter).</p>
<p>If you do decide to try it out, though, be sure you write your workflow so that it</p>
<pre><code class="lang-auto">uses: eps1lon/actions-label-merge-conflict@v2.0.1
</code></pre>
<p>instead of the version 2.0.0 that’s listed as the latest release at the project page. 2.0.0 has a bit of a retry-timeout-cascade problem thing that predictably hit us on our very first outing with it, so that I had to stop the workflow after a solid 30 minutes of spinning its wheels fruitlessly. 2.0.1 generally completes in under 5-10 minutes.</p>
<p>(There’s some expected lag to the whole process, because Github’s conflict-detection jobs are processed asynchronously… so any tool that acts on them has to be prepared to wait for those results to show up, after there’s activity on the branch.)</p>

**Comment by valentijnscholten** [comment]
<p>Thanks for the info and actions, they can be helpful. But in the end it’s a second best solution, might be better to have the green tickbox on the PR on the GitHub UI also turning red. It now stays green.</p>

**Comment by ferdnyc** [comment]
<p><a class="mention" href="/valentijnscholten">@valentijnscholten</a> Changing the color of the icon without changing the shape would impair the interface accessibility, as it would be undetectable to anyone with any sort of red-green colorblindness (the most common types).</p>
<p>Also, conflicts are independent of CI passing — IOW, you’d want to know when there are conflicts even if the CI status is a red X (CI failed) or a yellow bullet (CI in progress).</p>

**Comment by valentijnscholten** [comment]
<p>Maybe I meant that it should become a red cross or something else red. Even if the CI passes, the PR is not ready to merge due to the merge conflicts. They have to be resolved first en probably CI has to run again before it can be merged. So any sort of indicator that the PR is conflicted is helpful. And maybe in my mind the “have no conflicts” is mandatory to conclude a PR has “passed CI succesfully”. For now I have resorted to the GibHub Action promoted by GitHub to label conflicted PRs on every merge to dev.</p>

**Comment by jyn514** [comment]
How can I filter for PRs with conflicts in the GraphQL API? The query @francisfuzz posted will show the mergeable status, but I haven't figure out how to filter by it in the initial query. I want to search in rust-lang/rust, which has 600+ open PRs and hits the rate limits on the query explorer.

**Comment by jyn514** [comment]
It looks like the REST API doesn't return the mergeable status, so GraphQL it the only way to do this today.

**Comment by sylvainfilteau** [comment]
With github cli (`gh`) and `jq`, this outputs as a csv the "un-mergeable" pull requests from the last 100 pull requests of `repo` in `org`.

````bash
gh api graphql -F owner=org -F name=repo -f query='query($name: String!, $owner: String!) {
repository(owner: $owner, name: $name) {
    pullRequests(last: 100, states: OPEN) {
      edges {
        node {
          title
          number
          url
          author {
                login
          }
          mergeable
          state
        }
      }
    }
}
}' | jq -r '.data.repository.pullRequests.edges[] | select(.node.mergeable != "MERGEABLE") | [.node.author.login, .node.title, .node.url] | @csv' | sort
````

**Reply by jyn514** [reply]
That still hits the 100-PR rate limit, though.

**Reply by ferdnyc** [reply]
@jyn514 You can use pagination to overcome that:
```sh
gh api graphql --paginate -F owner=rust-lang -F name=rust -f query='
query($endCursor: String, $name: String!, $owner: String!) {
  repository(owner: $owner, name: $name) {
      pullRequests(first: 100, states: OPEN, after: $endCursor) {
        edges {
          node {
            title
            number
            url
            author {
                login
            }
            mergeable
            state
          }
        }
        pageInfo {
          hasNextPage
          endCursor
} } } }'
```
Will list all open PRs for rust-lang/rust, not just the first 100.

**Reply by ferdnyc** [reply]
(They'll come out in oldest-first order, due to `first: 100` being the starting point. Trying the same query with `last: 100` breaks the pagination. I don't know if you can reverse-paginate a query starting from the other end. I can't find any _obvious_ way to do it.)

**Comment by jgoedde** [comment]
Also in case anyone is looking for a git alias, I made it like this now using `gh` CLI similar to other posters.

```gitconfig
[alias]
    conflicted-prs = "!gh pr list --author \"@me\" \
      --base \"$(gh repo view --json defaultBranchRef --jq .defaultBranchRef.name)\" \
      --json number,title,mergeable,url \
      --jq '.[] | select(.mergeable == \"CONFLICTING\") | \"#\\(.number)  \\(.title)  \\(.url)\"'"
```
